pkg/activator/handler/concurrency_reporter_test.go

```diff
@@ -17,104 +17,299 @@ limitations under the License.
+						RequestCount:              1,
+						PodName:                   "activator",
+					},
+				},
+			},
 		},
-	}})
-
-	s.requestEnd(pod2)
-	s.requestEnd(pod1)
-
-	now = time.Now()
-	expectStats(t, s.report(now, 1), []*autoscaler.StatMessage{{
-		Key: pod1,
-		Stat: autoscaler.Stat{
-			Time:                      &now,
-			PodName:                   autoscaler.ActivatorPodName,
-			AverageConcurrentRequests: 1.0,
-			RequestCount:              0, // no new request arrived after reporting
+		{
+			name: "Multiple pods tick",
+			ops: []reqOp{
+				{
+					op:  requestOpStart,
+					key: "pod1",
+				},
+				{
+					op:  requestOpStart,
+					key: "pod2",
+				},
+				{
+					op: requestOpTick,
+				},
+				{
+					op:  requestOpEnd,
+					key: "pod1",
+				},
+				{
+					op:  requestOpStart,
+					key: "pod3",
+				},
+				{
+					op: requestOpTick,
+				},
+			},
+			expectedStats: []*autoscaler.StatMessage{
+				{
+					Key: "pod1",
+					Stat: autoscaler.Stat{
+						Time:                      &time.Time{},
+						AverageConcurrentRequests: 1,
+						RequestCount:              1,
+						PodName:                   "activator",
+					},
+				},
+				{
+					Key: "pod2",
+					Stat: autoscaler.Stat{
+						Time:                      &time.Time{},
+						AverageConcurrentRequests: 1,
+						RequestCount:              1,
+						PodName:                   "activator",
+					},
+				},
+				{
+					Key: "pod1",
+					Stat: autoscaler.Stat{
+						Time:                      &time.Time{},
+						AverageConcurrentRequests: 1,
+						RequestCount:              1,
+						PodName:                   "activator",
+					},
+				},
+				{
+					Key: "pod2",
+					Stat: autoscaler.Stat{
+						Time:                      &time.Time{},
+						AverageConcurrentRequests: 1,
+						RequestCount:              1,
+						PodName:                   "activator",
+					},
+				},
+				{
+					Key: "pod3",
+					Stat: autoscaler.Stat{
+						Time:                      &time.Time{},
+						AverageConcurrentRequests: 1,
+						RequestCount:              1,
+						PodName:                   "activator",
+					},
+				},
+				{
+					Key: "pod2",
+					Stat: autoscaler.Stat{
+						Time:                      &time.Time{},
+						AverageConcurrentRequests: 1,
+						RequestCount:              0,
+						PodName:                   "activator",
+					},
+				},
+				{
+					Key: "pod3",
+					Stat: autoscaler.Stat{
+						Time:                      &time.Time{},
+						AverageConcurrentRequests: 1,
+						RequestCount:              1,
+						PodName:                   "activator",
+					},
+				},
+			},
 		},
-	}})
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			writeDone := make(chan struct{})
+			writeWg := sync.WaitGroup{}
+			writeWg.Add(1)
+			s, cr := newTestStats(fakeClock{})
+			go func() {
+				cr.Run(writeDone)
+				writeWg.Done()
+			}()
+
+			readDone := make(chan struct{})
+			readWg := sync.WaitGroup{}
+			readWg.Add(1)
```

Comment:
I feel like all these WaitGroups and channels are quite complex and I don't see an immediate need to use them? 

I sketched a version without them here: https://github.com/knative/serving/commit/cdcf6b6a53742433fde663013c8b69454a527244 (ran successfully 10000 times with race detector)

Reply:
ah, yes - I didnt realize we were ok (and by default) making these channels buffered. Do you know what the motivation was for this (specifically statCh). ISTM this is fairly risky as its easy to be throwing messages in to a black hole without realizing it.

I'm fine pulling in this fix, but maybe we can loop back and see if were OK unbuffering these channels if possible?

Reply:
Personally, I tend to prefer buffered channels over having to involve more goroutines to make the test work. But sure, if you've got an alternative approach it'd be great to open a followup.